src/client/api/compile_check.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <gurt/debug.h>
+#include <daos.h>
+
+#if defined(__cplusplus)
+#define check_uuid(var, expect, type) (void)0
+#define check_string(var, expect, type) (void)0
+#else
+#define check_uuid(var, expect, type)							\
+	do {										\
+		printf("uuid check:   "#var" of type " #type " should be a %s, is %s\n",\
+		       expect ? "uuid" : "string", d_is_uuid(var) ? "uuid" : "string");	\
+		D_CASSERT(d_is_uuid(var) == expect);					\
+	} while (0)
+
+#define check_string(var, expect, type)								\
+	do {											\
+		printf("string check: "#var" of type " #type " should be a %s, is %s\n",	\
+		       expect ? "string" : "uuid", d_is_string(var) ? "string" : "uuid");	\
+		D_CASSERT(d_is_string(var) == expect);						\
+	} while (0)
+#endif
+
+#define noop
+
+#define FOREACH_UUID_TYPE(ACTION, poh, coh)				\
+	ACTION(uuid_t,       uuid,   noop, {0},  poh, coh, true)	\
+	ACTION(const uuid_t, cuuid,  noop, {0},  poh, coh, true)
+
+#define FOREACH_TYPE(ACTION, poh, coh)					\
+	FOREACH_UUID_TYPE(ACTION, poh, coh)				\
+	ACTION(char *,       charp,  noop, NULL, poh, coh, false)	\
+	ACTION(const char *, ccharp, noop, NULL, poh, coh, false)	\
+	ACTION(char,         arr,    [10], {0},  poh, coh, false)	\
+	ACTION(const char,   carr,   [10], {0},  poh, coh, false)	\
+	ACTION(char,         arru,   [37], {0},  poh, coh, false)	\
+	ACTION(const char,   carru,  [37], {0},  poh, coh, false)	\
+	ACTION(char,         dyn,    [],   "c",  poh, coh, false)
+
+#define FOREACH_ALL(ACTION, poh, coh)					\
+	FOREACH_TYPE(ACTION, poh, coh)					\
+	ACTION(literal,      "STR",  noop, noop, poh, coh, false)
+
+#define DECLARE_ACTION(type, name, mod, init, poh, coh, expect)	\
+	type	name mod = init;
+
+#define RUN_FUNCTION(cmd)	\
+	do {			\
+		int	__rc;	\
+		__rc = cmd;	\
+		(void)__rc;	\
+	} while (0)
+
+#define RUN_ACTION(type, name, mod, init, poh, coh, expect)			\
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon